opendut-util/opendut-auth/src/registration/client.rs

```diff
@@ -126,4 +136,92 @@ impl RegistrationClient {
             }
         }
     }
+    
+    pub async fn list_clients(&self, resource_id: Id) -> Result<Clients, RegistrationClientError> {
+        let access_token = self.inner.get_token().await
+            .map_err(|error| RegistrationClientError::RequestError { error: error.to_string(), cause: Box::new(error) })?;
+
+        let issuer_remote_url = request_uri(&self.config.issuer_remote_url, None)?;
+        let request = create_http_request_client(&access_token, &issuer_remote_url, http::Method::GET)?;
+        
+        let response = self.inner.reqwest_client.async_http_client(request)
+            .await;
+        match response { 
+            Ok(response) => {
+                 let clients: Clients = serde_json::from_slice(&response.body).unwrap();
+                 let filtered_clients = clients.0.into_iter().filter(|client| client.base_url.clone().is_some_and(|url| url.contains(&resource_id.value().to_string()))).collect::<Vec<Client>>();
+    
+                 Ok(Clients(filtered_clients))
+            }
+            Err(error) => {
+                 Err(RegistrationClientError::RequestError { error: "OIDC client list request failed!".to_string(), cause: Box::new(error) })
+            }
+        }
+    }
+
+    pub async fn delete_client(&self, resource_id: Id) -> Result<Clients, RegistrationClientError> {
+        let access_token = self.inner.get_token().await
+            .map_err(|error| RegistrationClientError::RequestError { error: error.to_string(), cause: Box::new(error) })?;
+
+        let clients = self.list_clients(resource_id).await?;
+        
+        let mut failed_deletion_clients = Vec::new();
+        
+        for client in clients.clone().0 {
+            let delete_client_url = request_uri(&self.config.issuer_remote_url, Some(client.client_id.to_string()))?;
+
+            let request = create_http_request_client(&access_token, &delete_client_url, http::Method::DELETE)?;
+
+            let response = self.inner.reqwest_client.async_http_client(request)
+                .await;
+
+            if response.is_err() {
+                failed_deletion_clients.push(client.client_id);
+            }
+        }
+        
+        if failed_deletion_clients.is_empty() {
+            Ok(clients)
+        } else {
+            Err( RegistrationClientError::ClientDeletionError { client_ids: failed_deletion_clients.join(",") } )
+        }
+    }
+}
+
+fn request_uri(issuer_remote_url: &Url, client_id: Option<String>) -> Result<Url, RegistrationClientError> {
```

Comment:
The name of the method does not reflect what is done here.

Reply:
Done

Reply:
The method did create the Keycloak Administrative API URI to either list all clients or a specific client. I moved the method inside the callee and added the Keycloak Admin API endpoint to the list of configuration parameters as this one was missing.

Changing the configuration is sadly a breaking change and will require a change in all deployments (add required environment variable).